pkg/templates/toolchaincluster/template.go

```diff
@@ -0,0 +1,60 @@
+package toolchaincluster
+
+import (
+	"embed"
+
+	v1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// GetServiceAccountFromTemplate returns the toolchaincluster service account object from the template content
+func GetServiceAccountFromTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (*v1.ServiceAccount, error) {
+	obj, gkv, err := decodeTemplate(toolchainclusterFS, templateName, into)
+	if err != nil {
+		return nil, err
+	}
+	if gkv.Kind == "ServiceAccount" {
+		return obj.(*v1.ServiceAccount), nil
+	}
+	return nil, nil
+}
+
+// GetRoleFromTemplate returns the toolchaincluster Role object from the template content
+func GetRoleFromTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (*rbac.Role, error) {
+	obj, gkv, err := decodeTemplate(toolchainclusterFS, templateName, into)
+	if err != nil {
+		return nil, err
+	}
+	if gkv.Kind == "Role" {
+		return obj.(*rbac.Role), nil
+	}
+	return nil, nil
+}
+
+// GetRoleBindingFromTemplate returns the toolchaincluster RoleBinding object from the template content
+func GetRoleBindingFromTemplate(toolchainclusterFS *embed.FS, templateName string, into runtime.Object) (*rbac.RoleBinding, error) {
```

Comment:
why do we actually have to care about the particular resources? Couldn't we just simply apply all resources no matter of their kinds? 
The only thing we have to rely on is the presence of the ServiceAccount to be able to get the needed token, but that's all.

Reply:
> why do we actually have to care about the particular resources? Couldn't we just simply apply all resources no matter of their kinds?

I'm not sure I'm following here, this logic is just converting the manifest yaml file that is embeded in the binary to a Golang type ( in this case is a RoleBinding ) . We will need the logic ( a controller I guess ) that will use all those functions and will actually apply/create those objects in the cluster. So it will need to know the kinds of the objects IMO.

Reply:
The alternative would be to just read the file content and apply it as is, without any sort of content validation, which in this case is done by checking the type of object we expect in the manifest.

Reply:
Also do we really think we will have other objects to deploy other than Role, RoleBinding and SA ? maybe ClusterRole and ClusterRoleBinding.

Reply:
If by content validation you're referring to decoding the template to get the object and gkv then I think that's only validating that it's a runtime object which is still generic? So that can be done the same way for all the objects and they can still be created/updated in a generic way.

I'm not sure that casting the object is equivalent to validating the object. If the template has some field that doesn't exist in the type then does it return an error or it's just ignored?

Reply:
> I'm not sure that casting the object is equivalent to validating the object. If the template has some field that doesn't exist in the type then does it return an error or it's just ignored?

That's true, it get's ignored, but you can unit test and check that required/expected fields are there. With generic types ( I believe you are referring to `runtime.Object` ) you can't check that as far as I know.

Reply:
You can use `unstructured.Unstructured{}` instead, there is 
```
runtime.DefaultUnstructuredConverter.FromUnstructured
```
that should convert the type into the specific object if needed